solr/core/src/java/org/apache/solr/filestore/NodeFileStore.java

```diff
@@ -0,0 +1,211 @@
+import org.apache.solr.api.JerseyResource;
+import org.apache.solr.client.api.endpoint.NodeFileStoreApis;
+import org.apache.solr.client.api.model.FileStoreDirectoryListingResponse;
+import org.apache.solr.client.api.model.FileStoreEntryMetadata;
+import org.apache.solr.client.api.model.FileStoreJsonFileResponse;
+import org.apache.solr.client.api.model.SolrJerseyResponse;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.params.CommonParams;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.apache.solr.common.params.SolrParams;
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.core.SolrCore;
+import org.apache.solr.jersey.PermissionName;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Implementation for {@link NodeFileStoreApis} */
+public class NodeFileStore extends JerseyResource implements NodeFileStoreApis {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private final CoreContainer coreContainer;
+  private final SolrQueryRequest req;
+  private final SolrQueryResponse rsp;
+  private final FileStore fileStore;
+
+  @Inject
+  public NodeFileStore(
+      CoreContainer coreContainer,
+      DistribFileStore fileStore,
+      SolrQueryRequest req,
+      SolrQueryResponse rsp) {
+    this.coreContainer = coreContainer;
+    this.req = req;
+    this.rsp = rsp;
+    this.fileStore = fileStore;
+  }
+
+  // TODO - this single "get" operation actually supports several different chunks of functionality:
+  //  syncing, directory listing, file-fetching, metadata-fetching. We should split it up into
+  //  multiple distinct APIs
+  @Override
+  @PermissionName(FILESTORE_READ_PERM)
+  public SolrJerseyResponse getFile(String path, Boolean sync, String getFrom, Boolean meta) {
+    final var response = instantiateJerseyResponse(SolrJerseyResponse.class);
+
+    if (Boolean.TRUE.equals(sync)) {
+      try {
+        fileStore.syncToAllNodes(path);
+        return response;
+      } catch (IOException e) {
+        throw new SolrException(SolrException.ErrorCode.BAD_REQUEST, "Error getting file ", e);
+      }
+    }
+
+    final var pathCopy = path;
+    if (getFrom != null) {
+      coreContainer
+          .getUpdateShardHandler()
+          .getUpdateExecutor()
+          .submit(
+              () -> {
+                log.debug("Downloading file {}", pathCopy);
+                try {
+                  fileStore.fetch(pathCopy, getFrom);
+                } catch (Exception e) {
+                  log.error("Failed to download file: {}", pathCopy, e);
+                }
+                log.info("downloaded file: {}", pathCopy);
+              });
+      return response;
+    }
+
+    if (path == null) {
+      path = "";
+    }
+    FileStore.FileType typ = fileStore.getType(path, false);
```

Comment:
"type"?

Reply:
This is something that carried over from the existing code, and I'm leery of getting sucked into larger improvements to that code.  There's a lot that could be done there.
But I can change it since it's a small modification.